pyro/contrib/funsor/enum_messenger.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2017-2019 Uber Technologies, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from collections import OrderedDict
+
+from pyro.poutine import Messenger
+from pyro.poutine.trace_messenger import TraceMessenger
+
+from pyro.contrib.funsor import named, to_funsor, to_data
+
+
+class SimpleEnumMessenger(Messenger):
+
+    def _pyro_sample(self, msg):
+
+        import funsor
+
+        if msg["done"] or msg["is_observed"] or msg.get("expand", False) or \
+                msg["infer"].get("enumerate") != "parallel":
+            return
+
+        msg["infer"]["funsor_fn"] = to_funsor(msg["fn"])
```

Comment:
Calling `to_funsor` in inference code on a `torch.Tensor` or `TorchDistribution` will automatically assign the correct names to all batch dimensions.